PullRequest for Issue103: Using signal map for AMDS scaler to identify the step scan finished

XuetaoLiu-S opened the pull request:
@davidChevrier @dretrex @iainworkman 

Hi guys, pls take a look at this PR. This supposed to fix the issues. I'm running the test right now.

There are also code refactors required for this. There are some redundant codes there to be optimized. I will do that in a sub branch.

XuetaoLiu-S committed:
Using signal map for AMDS scaler to identify the step scan finished

XuetaoLiu-S commented:
No rush... the step scan failed...

XuetaoLiu-S commented:
Scan is running now with restarting Acquaman

dretrex commented:
Once the scans are working a bit better, post in the thread again.  Then we can go over the completed code.

dretrex commented:
Does that mean we are good for review?

XuetaoLiu-S commented on source/Detector/Scaler/AMDSScalerDetector.cpp:
```diff
@@ -265,11 +269,79 @@ void AMDSScalerDetector::onTriggerDwellTimerTimeout()
+			// set the mapping of value changed. NOTE: This should be in a more graceful way
+			waitingChannels_.clear();
+
+			QMap<quint8, int> channelToIndexMap;
```
This section of codes are stolen from the `setFlattenedData()`. 

It will be optimized later

XuetaoLiu-S commented:
@dretrex yes, u can review now

dretrex commented on source/Detector/Scaler/AMDSScalerDetector.cpp:
```diff
@@ -229,25 +233,25 @@ void AMDSScalerDetector::onFetchScanBuffer()
+{
+	Q_UNUSED(value)
+
+	if(triggerDwellInterfaceDwellModeControl_->withinTolerance(1.0) && triggerDwellInterfaceStartControl_->withinTolerance(1.0)) {
```
Just wanting to double check that these are both supposed to be at 1 if we are wanting to stop acquisition?

dretrex commented on source/Detector/Scaler/AMDSScalerDetector.cpp:
```diff
@@ -229,25 +233,25 @@ void AMDSScalerDetector::onFetchScanBuffer()
-			}
+			startScalerDwellAcquisition();
 		}
 		else if(triggerDwellInterfaceStartControl_->withinTolerance(0.0) && triggerDwellInterfaceDwellModeControl_->withinTolerance(1.0) && triggerDwellTimer_->isActive()){
```
Is this supposed to be different from the line where my earlier comment is?

dretrex commented on source/Detector/Scaler/AMDSScalerDetector.cpp:
```diff
@@ -265,11 +269,79 @@ void AMDSScalerDetector::onTriggerDwellTimerTimeout()
+void AMDSScalerDetector::startScalerDwellAcquisition()
+{
+	double seconds = triggerDwellInterfaceDwellTimeControl_->value();
+	if(seconds > 0 && seconds < 100) {
```
Is there a reason for the `< 100` part of the conditional?

dretrex commented on source/Detector/Scaler/AMDSScalerDetector.cpp:
```diff
@@ -265,11 +269,79 @@ void AMDSScalerDetector::onTriggerDwellTimerTimeout()
+
+void AMDSScalerDetector::onScalerChannelValueChanged(int channelIndex)
+{
+	if(waitingChannels_.contains(channelIndex)){
```
Does this function only remove channels?  If so, I think this function should be renamed.

dretrex commented:
Just a couple comments.
Time Estimate: 30 minutes

XuetaoLiu-S committed:
rename AMDSThreadedScalerDetectorManager to AMDSThreadedScalerDetector

XuetaoLiu-S committed:
refactor the AMDSScalerDetector to two class: AMDSScalerDetctor (physical scaler) and AMDSDwellScalerDtector (softIOC AMDS scaler), added new AMDSScalerDetectorManager to manage the scaler detectors

XuetaoLiu-S committed:
add more comments

dretrex commented:
Overall, this still looks good.  Only thing now is to really make sure that we put it through its paces when we test it.

XuetaoLiu-S committed:
Update AMDSDwellScalerDetector.cpp

Make the helper functions const

XuetaoLiu-S committed:
Update AMDSDwellScalerDetector.h

Make helper functions const

XuetaoLiu-S committed:
- Fix PV connection bugs
- Fix the hard coded PV name in the setFlattenData() function
- A little bit code refactoring

dretrex commented:
@XuetaoLiu-CLS, those were the wrong const uses.  They are const functions: `bool isAcquiring() const;` not `const bool isAcquiring();`

XuetaoLiu-S committed:
Fix typo and the usage of const
